ecr-login/cache/file.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package cache
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	log "github.com/cihub/seelog"
+)
+
+const registryCacheVersion = "1.0"
+
+type RegistryCache struct {
+	Registries map[string]*AuthEntry
+	Version    string
+}
+
+type fileCredentialCache struct {
+	filename       string
+	cachePrefixKey string
+}
+
+func newRegistryCache() *RegistryCache {
+	return &RegistryCache{
+		Registries: make(map[string]*AuthEntry),
+		Version:    registryCacheVersion,
+	}
+}
+
+func NewFileCredentialsCache(filename string, cachePrefixKey string) CredentialsCache {
+	return &fileCredentialCache{filename: filename, cachePrefixKey: cachePrefixKey}
+}
+
+func (f *fileCredentialCache) Get(registry string) *AuthEntry {
+	log.Debugf("Checking file cache for %s", registry)
+	registryCache, err := f.load()
+	if err != nil {
+		log.Infof("Could not load existing cache: %v", err)
+		f.Clear()
+		registryCache = newRegistryCache()
+	}
+	return registryCache.Registries[f.cachePrefixKey+registry]
+}
+
+func (f *fileCredentialCache) Set(registry string, entry *AuthEntry) {
+	log.Debugf("Saving credentials to file cache for %s", registry)
+	registryCache, err := f.load()
+	if err != nil {
+		log.Infof("Could not load existing cache: %v", err)
+		f.Clear()
+		registryCache = newRegistryCache()
+	}
+
+	registryCache.Registries[f.cachePrefixKey+registry] = entry
+
+	err = f.save(registryCache)
+	if err != nil {
+		log.Infof("Could not save cache: %s", err)
+	}
+}
+
+func (f *fileCredentialCache) Clear() {
+	err := os.Remove(f.filename)
+	if err != nil {
+		log.Infof("Could not clear cache: %s")
+	}
+}
+
+// Saves credential cache to disk. This writes to a temporary file first, then moves the file to the config location.
+// This elminates from reading partially written credential files, and reduces (but does not eliminate) concurrent
+// file access. There is not guarantee here for handling multiple writes at once since there is no out of process locking.
+func (f *fileCredentialCache) save(registryCache *RegistryCache) error {
+	defer log.Flush()
+	file, err := ioutil.TempFile("", "ecr-credential-cache")
```

Comment:
Can you change this to use a temp file in the home directory?  `os.Rename` will fail if the temp file and destination file are not on the same filesystem, so creating the temp file in the same directory would help that problem.

```go
dir, _ := homedir.Dir()
file, err := ioutil.TempFile(dir, "ecr-credential-cache")
```